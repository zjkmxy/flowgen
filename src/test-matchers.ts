import { execFileSync } from "child_process";

import chalk from "chalk";
import flow from "flow-bin";

import { beautify } from ".";

// eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars
declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace jest {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars,no-unused-vars
    interface Matchers<R> {
      toBeValidFlowTypeDeclarations(): R;
    }
  }
}

expect.extend({
  async toBeValidFlowTypeDeclarations(source) {
    const beautifiedSource = await beautify(source);
    try {
      execFileSync(
        flow,
        ["check-contents", "--color=always", "--timeout=30"],
        {
          input: beautifiedSource,
          stdio: ["pipe", "pipe", "pipe"],
        },
      );
    } catch (err) {
      return {
        message: () =>
          `expected ${chalk.bold(
            beautifiedSource.trimEnd(),
          )} to be valid flow:\n${chalk.red(err.stdout)}`,
        pass: false,
      };
    }

    return {
      message: () =>
        `expected ${chalk.bold(
          beautifiedSource.trimEnd(),
        )} not to be valid flow`,
      pass: true,
    };
  },
});
