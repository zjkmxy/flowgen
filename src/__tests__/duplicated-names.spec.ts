import { compiler, beautify } from "..";
import "../test-matchers";

it("should handle variable & type having same name", async () => {
  const ts = `
  export declare const AuthMechanism: {
    readonly MONGODB_AWS: "MONGODB-AWS";
    readonly MONGODB_CR: "MONGODB-CR";
  };
  export declare type AuthMechanism = typeof AuthMechanism[keyof typeof AuthMechanism];
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should generate unique names", async () => {
  const ts = `
  export declare type AuthMechanismType = string;
  export declare const AuthMechanism: {
    readonly MONGODB_AWS: "MONGODB-AWS";
    readonly MONGODB_CR: "MONGODB-CR";
  };
  export declare type AuthMechanism = typeof AuthMechanism[keyof typeof AuthMechanism];
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});

it("should be scoped to main file", async () => {
  const ts = `
  import { Buffer } from 'buffer';
  export declare type BufferAlias = Buffer;
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  // FlowJs is not valid due to `import` not transpiled to typeof
  // expect(result).toBeValidFlowTypeDeclarations();
});

it("should support generic type rename", async () => {
  const ts = `
export declare const ProfilingLevel: Readonly<{
  readonly off: "off";
}>;
export declare type ProfilingLevel = typeof ProfilingLevel[keyof typeof ProfilingLevel];

export declare type Callback<T = any> = (error?: Error, result?: T) => void;

export declare const callback: Callback<ProfilingLevel>;
`;
  const result = compiler.compileDefinitionString(ts, { quiet: true });
  expect(await beautify(result)).toMatchSnapshot();
  expect(result).toBeValidFlowTypeDeclarations();
});
