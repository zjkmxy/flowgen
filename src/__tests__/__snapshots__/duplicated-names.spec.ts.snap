// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be scoped to main file 1`] = `
"import { Buffer } from "buffer";
export type BufferAlias = Buffer;
"
`;

exports[`should generate unique names 1`] = `
"export type AuthMechanismType = string;
declare export var AuthMechanism: {
  +MONGODB_AWS: "MONGODB-AWS",
  +MONGODB_CR: "MONGODB-CR",
};
export type AuthMechanismType1 =
  (typeof AuthMechanism)[keyof typeof AuthMechanism];
"
`;

exports[`should handle variable & type having same name 1`] = `
"declare export var AuthMechanism: {
  +MONGODB_AWS: "MONGODB-AWS",
  +MONGODB_CR: "MONGODB-CR",
};
export type AuthMechanismType =
  (typeof AuthMechanism)[keyof typeof AuthMechanism];
"
`;

exports[`should support generic type rename 1`] = `
"declare export var ProfilingLevel: $ReadOnly<{
  +off: "off",
}>;
export type ProfilingLevelType =
  (typeof ProfilingLevel)[keyof typeof ProfilingLevel];
export type Callback<T = any> = (error?: Error, result?: T) => void;
declare export var callback: Callback<ProfilingLevelType>;
"
`;
